librbd: add a crypto object dispatch layer

orozery opened the pull request:
This is my first PR, so apologies in advance :)

This is a first in a series of PR's targeted to enable encryption of RBD images.
The encryption/decryption function of an object extent will depend on:
1. An encryption key, unique per image
2. The logical image offset (this will enable supporting encryption modes such as XTS)

This PR assumes a non-aligned encryption mode, which is simple to support but obviously not practical.

The next couple of PRs will enable encryption alignment (specifically required for doing aligned writes), and will include adding the following APIs to the ObjectDispatchInterface:
1. Scatter-gather reading of object extents + returning an object version
2. Writing of an object extent while asserting a previous version

Later on, we will also make the encryption layer compatible with cloning, by:
1. re-factoring object copyup to go through the object dispatch layers while writing
2. add a flag to the object ReadResult indicating that the returned bufferlist is encrypted/decrypted (since reading from the parent image automatically decrypts data)

This will conclude that data-path handling.

In parallel, a colleague of mine is working on the control path, responsible for initializing the encryptor/decryptor on an image create/open operation, as well as persisting required crypto metadata to the image header.

orozery commented:
@dillaman can you please add an rbd label?

dillaman commented on src/librbd/ImageCtx.h:
```diff
@@ -225,6 +226,8 @@ namespace librbd {
     exclusive_lock::Policy *exclusive_lock_policy = nullptr;
     journal::Policy *journal_policy = nullptr;
 
+    crypto::CryptoInterface *crypto = nullptr;
```
Nit: totally fine for now, but we will eventually want to move the core crypto dispatch logic to a librbd plug-in like the parent object cache.

orozery replied:
Makes sense. I will re-factor this PR as a plugin

dillaman commented on src/test/librbd/mock/MockImageCtx.h:
```diff
@@ -303,6 +304,7 @@ struct MockImageCtx {
   MockObjectMap *object_map;
   MockExclusiveLock *exclusive_lock;
   MockJournal *journal;
+  crypto::MockCryptoInterface *crypto;
```
```suggestion
  crypto::MockCryptoInterface *crypto = nullptr;
```

dillaman commented on src/test/librbd/mock/MockImageCtx.h:
```diff
@@ -91,7 +92,7 @@ struct MockImageCtx {
       state(new MockImageState()),
       image_watcher(NULL), object_map(NULL),
-      exclusive_lock(NULL), journal(NULL),
+      exclusive_lock(NULL), journal(NULL), crypto(NULL),
```
```suggestion
      exclusive_lock(NULL), journal(NULL),
```

dillaman commented on src/librbd/crypto/CryptoObjectDispatch.cc:
```diff
@@ -0,0 +1,197 @@
+                 << object_off << "~" << object_len << dendl;
+
+  *dispatch_result = io::DISPATCH_RESULT_COMPLETE;
+  auto req = new C_EncryptedObjectReadRequest(
```
Why not just let the object request proceed to the next layer and intercept the result for decryption? You can wrap the provided `on_finish` handler so that you can get a callback when it finishes the read. It won't work for the parent-read case, but neither will this approach currently.

orozery replied:
Well I have already written it that way to prepare for alignment, where you need to modify the original read request.
Since I already have this code, I prefer to avoid writing throw-away code.

dillaman replied:
All the ops will need to handle alignment crypto block alignment, correct? write, write-same, and compare-and-write might need to perform a read-modify-write cycle. And discard would actually need to write encrypted zeroes(?) or perhaps just turn it into a no-op if you are going to store integrity.

orozery replied:
Yep.
I actually mistaken Discard to be a cache operation, but I now see that it actually is writing zeros, right? (hopefully claiming back OSD space, right?)

dillaman replied:
It's essentially a hint back to the OSD to reclaim space if possible. We (librbd) try to avoid issuing an actual "zero"-ing write operation on a discard and instead just treat the discard as a hint and ignore any extents that will cause us not to reclaim space.

dillaman commented on src/librbd/crypto/CryptoObjectDispatch.cc:
```diff
@@ -0,0 +1,197 @@
+                 << dendl;
+
+  uint64_t image_offset = get_image_offset(m_image_ctx, object_no, object_off);
+  m_image_ctx->crypto->encrypt(std::move(cmp_data), image_offset);
```
Nit: this is going to pose an interesting challenge for the unique IV mode. You would need to encrypt the `cmp_data` with the same key+IV as it was previously encoded with.

orozery replied:
You're right.
Moreover, even without unique IV you may still need to read data before you encrypt cmp_data, for alignment reasons.

dillaman commented on src/librbd/crypto/CryptoObjectDispatch.cc:
```diff
@@ -0,0 +1,197 @@
+}
+
+template <typename I>
+uint64_t CryptoObjectDispatch<I>::get_image_offset(
```
If this is attempting to perform full-image LUKS-like encryption, perhaps there really should be two different crypto plugins instead of trying to fit them both into the object-level encryption solution.

orozery replied:
CryptoInterface seems generic enough to me, as I assume most of not all realistic encryption modes will require the image offset as a parameter.

orozery replied:
But you're right about that from a cryptographic point of view, it should not matter if we replace the image offset by some 1-to-1 mapping like (object_no, object_off).

dillaman commented on src/librbd/crypto/CryptoObjectDispatch.h:
```diff
@@ -0,0 +1,104 @@
+          uint64_t journal_tid, uint64_t new_journal_tid) override {
+  }
+
+  static uint64_t get_image_offset(
```
Nit: perhaps just move this as a helper in `osdc/Striper` to avoid the need for a helper function here. It would certainly be lighter weight than creating a vector for extents.

orozery replied:
Good idea. I will open up a seperate `osdc\Striper` PR to add that

orozery commented:
Changes pushed:
1. Added Discard handling by converting to write_same of zeros.
2. Moved CryptoInterface member from ImageCtx to CryptoObjectDispatch
3. Refactored unit-test with SetUp and TearDown

orozery commented:
rebased + switched to Striper::get_file_offset

dillaman commented:
Looks like this has a build failure:

https://jenkins.ceph.com/job/ceph-dev-new-build/ARCH=x86_64,AVAILABLE_ARCH=x86_64,AVAILABLE_DIST=centos8,DIST=centos8,MACHINE_SIZE=gigantic/43993//consoleFull

dillaman commented:
jenkins test make check

orozery commented:
rebased and fixed build error

dillaman commented:
This passes the jenkins build but fails on the real builders: https://jenkins.ceph.com/job/ceph-dev-new-build/ARCH=x86_64,AVAILABLE_ARCH=x86_64,AVAILABLE_DIST=centos8,DIST=centos8,MACHINE_SIZE=gigantic/44120/console

orozery commented:
how can I re-run this test and get the matching jenkins link?

dillaman commented:
> how can I re-run this test and get the matching jenkins link?

It's not a test -- it's just the build process. I think the version of GCC just couldn't determine that `C_EncryptedObjectReadRequest` should have a template type of `I` -- so perhaps just tweak it to be explicit? Also needs to be rebased due to the merge of your read extent PR.

dillaman commented:
jenkins test make check
